src/group_theory/free_product.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import algebra.free_monoid
+import group_theory.congruence
+/-!
+# The free product of groups or monoids
+Given an `ι`-indexed family `M` of monoids, we define their free product (categorical coproduct)
+`free_product M`.
+
+When `M i` are all groups, `free_product M` is also a group (and the coproduct in the category of
+groups).
+
+The equivalence with reduced words should be added to this file soon.
+
+## Main definitions
+
+- `free_product M`: the free product, defined as a quotient of a free monoid.
+- `free_product.of {i} : M i →* free_product M`.
+- `free_product.lift : (Π {i}, M i →* N) ≃ (free_product M →* N)`: the universal property.
+
+## Remarks
+There are many answers to the question "what is the free product of a family `M` of monoids?", and
+they are all equivalent but not obviously equivalent. We provide two answers. The first, almost
+tautological answer is given by `free_product M`, which is a quotient of the type of words in the
+alphabet `Σ i, M i`. It's straightforward to define and easy to prove its universal property. But
+this answer is not completely satisfactory, because it's difficult to tell when two elements
+`x y : free_product M` are distinct since `free_product M` is defined as a quotient.
+
+The second, maximally efficient answer is given by `word M`. An element of `word M` is a word in the
```

Comment:
This paragraph refers to code that was split to a separate PR. Perhaps just move this paragraph to that PR?